README.md
# RPistepper

RPistepper is a library containing:
* A class to control a stepper motor with a RPi.
* A function to execute a zig-zag motion with two motors.
* A function to execute a square_spiral motion with two motors.

## Wiring
In our setup, the power to the motors (Vm) is supplied with the 5V pins of the RPi, the grounding of the coils is controlled with a [ULN2803A](http://www.ti.com/lit/ds/symlink/uln2803a.pdf) transistor array.

![Alt text](pinout.png "Example setup")

#### Conections RPi - ULN2803A:

| RPi Pin (BCM)| ULN2803A|
|--------------|---------|
|      17      |    1B   |
|      27      |    2B   |
|      10      |    3B   |
|      9       |    4B   |
|      14      |    5B   |
|      15      |    6B   |
|      23      |    7B   |
|      24      |    8B   |

#### Conections ULN2803A - Motors:

| ULN2803A| Motors          |
|---------|-----------------|
|    1C   | Motor_1 Coil_A1 |
|    2C   | Motor_1 Coil_A2 |
|    3C   | Motor_1 Coil_B1 |
|    4C   | Motor_1 Coil_B2 |
|    5C   | Motor_2 Coil_A1 |
|    6C   | Motor_2 Coil_A2 |
|    7C   | Motor_2 Coil_B1 |
|    8C   | Motor_2 Coil_B2 |

In this case, two motors were attached to the ULN2803A.


## Usage
### class Motor
This class allows the user to control a 6 pin stepper motor using 4 GPIO pins of a RPi.

Software uses BCM mode for pin indexing.

This class is best used with the 'with' statement to properly handle the cleanup of the GPIOs.

self.steps is a property of this class that will get the number of steps taken from the initial position or set to a specific step, similar to self.move.

In order to save power, it's advised to call self.release() when the motor is idle.

Arguments are a list with the 4 pins (Coil_A1, Coil_A2, Coil_B1, Coil_B2), the delay between steps (default = 20ms) and verbose to display reports on the motor movements, the last two are optional.
e.g:
```python
import RPistepper as stp
M1_pins = [17, 27, 10, 9]
with stp.Motor(M1_pins) as M1:
    for i in range(10):               # moves 20 steps,release and wait
        print M1
        M1.move(20)
        M1.release()
        raw_input('enter to execute next step')
```
If the class is instantiated normally, use the method `cleanup` prior to closing the application to close the GPIO resources.
Also, if it's important to go back to the initial position when finishing the routine, use the method `reset`.

```python
import RPistepper as stp
M1_pins = [17, 27, 10, 9]
M1 = stp.Motor(M1_pins)
for i in range(10):               # moves 20 steps,release and wait
    print M1
    M1.move(20)
    M1.release()
    raw_input('enter to execute next step')
M1.reset()
M1.cleanup()
```


#### Methods
Currently there are five implemented methods:
```python
def move(self, steps):
    '''
    Moves the motor 'steps' steps. Negative steps moves the motor backwards
    '''
```
```python
def release(self):
    '''
    Sets all pins low. Power saving mode
    '''
```
```python
def reset(self):
    '''
    Returns the motor to it's initial position
    '''
```
```python
def zero(self):
    '''
    Sets the motor to the next position which Coil_A1 and Coil_A2
    are on. Sets this position as the reference (steps = 0).
    '''
```
```python
def cleanup(self):
    '''
    Cleans the GPIO resources
    '''
```
The main method is `move`, which moves the motor the desired number of steps

#### steps property
It's possible to check the motor position or manually set the desired step using the `steps` property:
```python
import RPistepper as stp
M1_pins = [17, 27, 10, 9]
with stp.Motor(M1_pins) as M1:
    for i in range(10):               # moves 20 steps,release and wait
        print M1.steps
        M1.steps = 20*i
        M1.release()
        raw_input('enter to execute next step')
    M1.reset()
```
#### Attributes
This class haves the following attributes:

| Attribute    |  Data   |
|--------------|---------|
| DELAY        | Time between steps           |
| VERBOSE      | Display motor data on screen |
| PINS         | GPIOs used by the instance   |
| actual_state | A list with the status of the coils (on/off) |

### functions
These two functions executes pre determined movements and requires two stepper motor objects:
```python
def zig_zag(motor1, motor2, amp1, amp2, delay=None):
    '''
    Executes a zig-zag movement with two RPistepper objects.
    Arguments are: motor1 and motor2 objects and amp1, amp2, the amplitude
    of movement, a tuple (step, rep) representing the number of steps per
    iteration and the number of iterations of the following algorithm:
        Repeat rep1 times:
            1. Moves motor 2 step2*rep2 steps forward
            2. Moves motor 1 step1 steps forward
            3. Moves motor 2 step2*rep2 steps backwards
            4. Moves motor 1 step1 steps forward
        Reset to initial state
        Release the motors
    It's possible to change the delay between steps with the 'delay' argument
    '''
```
```python
def square_spiral(motor1, motor2, amplitude, delay=None):
    '''
    Executes a square spiral movement with two RPistepper objects.
    Arguments are: motor1 and motor2 objects and the amplitude of movement,
    a tuple (step, rep) representing the number of steps per iteration and
    the number of iterations of the following algorithm:
        for i in range(rep):
            1. Moves motor 2 to position i
            2. Moves motor 1 to position i
            3. Moves motor 1 to position -i
            4. Moves motor 2 to position -i
        Reset to initial state
        Release the motors
    It's possible to change the delay between steps with the 'delay' argument
    '''
```


CHANGELOG.md
# CHANGELOG

## 0.3a0
* Added zero method
* Changed `RPiStepper` class to `Motor` class
* Added RPistepper shell
* Added RPistepper gui
* Added `rpistepper` command
* in progress

## 0.2a0
* Fixed documentation
* GPIO output callings now in list/tuple format
* Fixed python 3 support

## 0.1a0
* Initial release


example.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
RPistepper example
'''
#______________________________________________________________________
# imports
import RPistepper as stp

#______________________________________________________________________
# main test
# Motor Pins: BCM
M1_pins = [17, 27, 10, 9]
M2_pins = [14, 15, 23, 24]
with stp.Motor(M1_pins) as M1, stp.Motor(M2_pins) as M2:
    M1.VERBOSE = True
    M2.VERBOSE = True
    #__________________________________________________________________
    # zig_zag
    stp.zig_zag(M1, M2, (5, 10), (5, 10))
    #__________________________________________________________________
    # square_spiral
    stp.square_spiral(M1, M2, (5, 10))
    #__________________________________________________________________
    # some movement
    for i in range(10): # repeat 10 times
        print(M1)       # show M1 data
        M1.move(20)     # move 20 steps
        print(M1)       # show M1 data
        M1.steps = -20  # move to position -40
        M1.zero()       # recalibrates the reference position
        print(M1)       # show M1 data
        M1.steps = 0    # move to position
    #__________________________________________________________________
    # reset motors to original position
    M1.reset()
    M2.reset()


RPistepper.py
    #!/usr/bin/env python2
# -*- coding: utf-8 -*-
'''
RPistepper is a library containing:
    * A class to control a stepper motor with a RPi.
    * A function to execute a zig-zag motion with two motors.
    * A function to execute a square_spiral motion with two motors.

Copyright (C) 2015 Luiz Eduardo Amaral <luizamaral306@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

#______________________________________________________________________
# imports
import RPi.GPIO as GPIO
from time import sleep

#______________________________________________________________________
# classezz
class RPiStepper(object):
    '''
    This class allows the user to control a 6 pin stepper motor using
    4 GPIO pins of a RPi.

    Software uses BCM mode for pin indexing.

    This class is best used with the 'with' statement to properly
    handle the cleanup of the GPIOs.

    self.steps is a property of this class that will get the number of
    steps taken from the initial position or set to a specific step,
    similar to self.move.

    In order to save power, it's advised to call self.release() when
    the motor is idle.
    '''

    #__________________________________________________________________
    # class attributes
    DELAY = 0.02
    VERBOSE = False

    #__________________________________________________________________
    # magic methods
    def __init__(self, pins, delay=DELAY, verbose=VERBOSE):
        '''
        Arguments are a list with the 4 pins (Coil_A1, Coil_A2,
        Coil_B1, Coil_B2), the delay between steps (default = 20ms) and
        verbose to display reports on the motor movements, the last two
        are optional.
        '''
        self.PINS = pins
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(self.PINS, GPIO.OUT)
        self.DELAY = delay
        self.VERBOSE = verbose
        self.actual_state = []
        self._step_list = [
            (1, 1, 0, 0),
            (1, 0, 1, 0),
            (0, 1, 1, 0),
            (0, 1, 0, 1),
            (0, 0, 1, 1),
            (1, 0, 0, 1)]
        self._zero = [0]*4
        self._steps = 0
        self._set_step(self._step_list[0])
        self.release()

    @property
    def steps(self):
        '''
        Number of steps taken from the initial position
        '''
        return self._steps

    @steps.setter
    def steps(self, value):
        steps = value - self._steps
        self.move(steps)

    def __repr__(self):
        return 'Motor at pins: {0}, Steps: {1}, Position: {2}'.format(
            self.PINS, self.steps, self.actual_state)

    def __enter__(self):
        return self

    def __exit__(self, type, value, tb):
        self.cleanup()

    #__________________________________________________________________
    # methods
    def move(self, steps):
        '''
        Moves the motor 'steps' steps. Negative steps moves the motor
        backwards
        '''
        if steps == 0:
            return
        if self.VERBOSE:
            print str(self)+ ', Moving: {0} steps'.format(steps)
        rotation = steps/abs(steps)
        for i in range(0, steps, rotation):
            index = (self._steps + rotation)%len(self._step_list)
            self._set_step(self._step_list[index])
            sleep(self.DELAY)
            self._steps += rotation

    def release(self):
        '''
        Sets all pins low. Power saving mode
        '''
        self._set_step(self._zero)
        self.actual_state = self._zero

    def reset(self):
        '''
        Returns the motor to it's initial position
        '''
        self.steps = 0

    def cleanup(self):
        '''
        Cleans the GPIO resources
        '''
        GPIO.cleanup(self.PINS)

    #__________________________________________________________________
    # private methods
    def _set_step(self, states):
        '''
        Sets the pins A_1, A_2, B_1, B_2, 1 or 0 (HIGH ou LOW)
        '''
        self.actual_state = states
        for pin, state in zip(self.PINS, states):
            GPIO.output(pin, state)

#______________________________________________________________________
# functions
def zig_zag(motor1, motor2, amp1, amp2, delay=None):
    '''
    Executes a zig-zag movement with two RPistepper objects.
    Arguments are: motor1 and motor2 objects and amp1, amp2, the
    amplitude of movement, a tuple (step, rep) representing the number
    of steps per iteration and the number of iterations of the following
    algorithm:
        Repeat rep1 times:
            1. Moves motor 2 step2*rep2 steps forward
            2. Moves motor 1 step1 steps forward
            3. Moves motor 2 step2*rep2 steps backwards
            4. Moves motor 1 step1 steps forward
        Reset to initial state
        Release the motors
    It's possible to change the delay between steps with the 'delay'
    argument
    '''
    step1, rep1 = amp1
    step2, rep2 = amp2
    if delay:
        motor1.delay = delay
        motor2.delay = delay
    for i in range(rep1):
        motor2.move(step2*rep2)
        motor1.move(step1)
        motor2.move(-step2*rep2)
        motor1.move(step1)
    motor1.reset()
    motor2.reset()
    motor1.release()
    motor2.release()

def square_spiral(motor1, motor2, amplitude, delay=None):
    '''
    Executes a square spiral movement with two RPistepper objects.
    Arguments are: motor1 and motor2 objects and the amplitude of movement,
    a tuple (step, rep) representing the number of steps per iteration and
    the number of iterations of the following algorithm:
        for i in range(rep):
            1. Moves motor 2 to position i
            2. Moves motor 1 to position i
            3. Moves motor 1 to position -i
            4. Moves motor 2 to position -i
        Reset to initial state
        Release the motors
    It's possible to change the delay between steps with the 'delay'
    argument
    '''
    step, rep = amplitude
    if delay:
        motor1.delay = delay
        motor2.delay = delay
    for i in range(rep):
        motor2.steps = (i+1)*step
        motor1.steps = (i+1)*step
        motor1.steps = -(i+1)*step
        motor2.steps = -(i+1)*step
    motor1.reset()
    motor2.reset()
    motor1.release()
    motor2.release()


RPistepper/gui.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#______________________________________________________________________
# imports
import RPistepper as stp
from sys import version_info
from collections import OrderedDict

if version_info[0] == 3:
    import tkinter as tk
    from tkinter import ttk
else:
    import Tkinter as tk
    from Tkinter import ttk

#______________________________________________________________________
# classes
class GUI(tk.Tk):
    PINS = OrderedDict()
    PINS['m0'] = stp.m0
    PINS['m1'] = stp.m1
    PINS['m2'] = stp.m2
    PINS['m3'] = stp.m3
    PINS['m4'] = stp.m4
    PINS['m5'] = stp.m5
    PINS['m6'] = stp.m6

    B_WIDTH = 15

    def __init__(self, *args, **kwargs):
        super(GUI, self).__init__(*args, **kwargs)
        self.title('RPiStepper GUI')

        self.mainframe = ttk.Frame(self, padding=3)
        self.mainframe.grid(column=0, row=0, sticky=(tk.N, tk.W, tk.E, tk.S))
        self.mainframe.rowconfigure(0, weight=1)
        self.mainframe.columnconfigure(0, weight=1)

        self.menubar = tk.Menu()

        self.filemenu = tk.Menu(self.menubar, tearoff=0)
        self.filemenu.add_command(label='Exit', command=lambda: self.quit())
        self.menubar.add_cascade(label='File', menu=self.filemenu)

        self.modemenu = tk.Menu(self.menubar, tearoff=0)
        self.modemenu.add_command(label='Control Motors',
            command=self.switch_control_motors_view)
        self.modemenu.add_command(label='2D Movements',
            command=self.switch_movements_2d_view)
        self.menubar.add_cascade(label='Mode', menu=self.modemenu)

        self.helpmenu = tk.Menu(self.menubar, tearoff=0)
        self.helpmenu.add_command(label='About', command=lambda:
            self.dialog_window('About RPistepper', 'RPistepper is an application to control steper motors with a Raspberry Pi'))
        self.menubar.add_cascade(label='Help', menu=self.helpmenu)

        self.config(menu=self.menubar)

        self.motor_block = OrderedDict()
        self.motor_object = OrderedDict()

        self.motor_label = OrderedDict()
        self.move_entry = OrderedDict()
        self.motor_button = OrderedDict()
        self.motor_status = OrderedDict()
        self.motor_status_var = OrderedDict()
        self.motor_on = OrderedDict()

        self.on_off_button = ttk.Frame(self.mainframe, border=1, relief=tk.SUNKEN)
        self.on_off_button.grid(column=0, row=0, sticky=(tk.N, tk.W, tk.E, tk.S),
            padx=3)
        self.motor_view = ttk.Frame(self.mainframe, border=1, relief=tk.SUNKEN)
        self.motor_view.grid(column=0, row=1, sticky=(tk.N, tk.W, tk.E, tk.S),
            padx=3)
        self.movements_2d_view = ttk.Frame(self.mainframe, border=1, relief=tk.SUNKEN)
        self.movements_2d_view.grid(column=0, row=0, sticky=(tk.N, tk.W, tk.E, tk.S),
            padx=3)

        for column, motor in enumerate(self.PINS):
            self.create_motor_block(motor, column)

        self.zig_zag_frame = {}
        self.zig_zag_frame['frame'] = ttk.Frame(self.movements_2d_view, relief=tk.SUNKEN, padding=10)
        self.create_function_frame(self.zig_zag_frame, 'Zig-Zag:\nChoose 2 motors')
        self.zig_zag_frame['frame'].grid(row=0, column=0)
        self.zig_zag_frame['button']['command'] = self.execute_zig_zag_cmd

        self.s_spiral_frame = {}
        self.s_spiral_frame['frame'] = ttk.Frame(self.movements_2d_view, relief=tk.SUNKEN, padding=10)
        self.create_function_frame(self.s_spiral_frame, 'Square-Spiral:\nChoose 2 motors')
        self.s_spiral_frame['frame'].grid(row=0, column=1, sticky=(tk.N, tk.S, tk.E, tk.W))
        self.s_spiral_frame['button']['command'] = self.execute_s_spiral_cmd
        self.s_spiral_frame['amp_y'].grid_remove()
        self.s_spiral_frame['rep_y'].grid_remove()
        self.s_spiral_frame['label_1y'].grid_remove()
        self.s_spiral_frame['label_2y'].grid_remove()
        self.s_spiral_frame['label_2x'].grid(row=8, column=1)
        self.s_spiral_frame['rep_x'].grid(row=9, column=1)
        self.s_spiral_frame['label_1x']['text'] = 'Amplitude'
        self.s_spiral_frame['label_2x']['text'] = 'Repetitions'

        self.switch_movements_2d_view()
        # self.switch_control_motors_view()

    def dialog_window(self, title, text):
        top = tk.Toplevel()
        top.title(title)

        msg = tk.Message(top, text=text)
        msg.grid()

        button = tk.Button(top, text='Dismiss', command=top.destroy)
        button.grid()

    def update_message(self, motor):
        steps = self.motor_object[motor].steps
        if self.motor_object[motor].locked:
            status = 'locked'
        else:
            status = 'released'
        self.motor_status_var[motor].set('Motor {0}\nsteps: {1}'.format(
            status, steps))

    def switch_control_motors_view(self):
        self.movements_2d_view.grid_remove()
        self.on_off_button.grid()
        self.motor_view.grid()

    def switch_movements_2d_view(self):
        self.movements_2d_view.grid()
        self.on_off_button.grid_remove()
        self.motor_view.grid_remove()

    def create_function_frame(self, frame_dict, title):
        frame_dict['label_0'] = tk.Label(frame_dict['frame'], text=title)
        frame_dict['label_0'].grid()

        frame_dict['checkbox_frame'] = ttk.Frame(frame_dict['frame'],
            relief=tk.SUNKEN)
        frame_dict['checkbox_frame'].grid()
        frame_dict['checkbox'] = {}

        frame_dict['label_x'] = tk.Label(frame_dict['checkbox_frame'], text='x axis')
        frame_dict['label_x'].grid(row=0, column=0)
        frame_dict['label_y'] = tk.Label(frame_dict['checkbox_frame'], text='y axis')
        frame_dict['label_y'].grid(row=0, column=1)
        frame_dict['checkbox']['var_x'] = tk.IntVar()
        frame_dict['checkbox']['var_y'] = tk.IntVar()
        frame_dict['checkbox']['var_x'].set(0)
        frame_dict['checkbox']['var_y'].set(1)
        for row, motor in enumerate(self.PINS):
            frame_dict['checkbox'][motor+'_x'] = ttk.Radiobutton(
                frame_dict['checkbox_frame'], text='motor '+motor, variable=frame_dict['checkbox']['var_x'], value=row)
            frame_dict['checkbox'][motor+'_x'].grid(row=row+1, column=0)
            frame_dict['checkbox'][motor+'_y'] = ttk.Radiobutton(
                frame_dict['checkbox_frame'], text='motor '+motor, variable=frame_dict['checkbox']['var_y'], value=row)
            frame_dict['checkbox'][motor+'_y'].grid(row=row+1, column = 1)

        frame_dict['label_1x'] = tk.Label(frame_dict['checkbox_frame'], text='Amp x')
        frame_dict['label_1x'].grid(row=row+2, column=0)

        frame_dict['label_1y'] = tk.Label(frame_dict['checkbox_frame'], text='Amp y')
        frame_dict['label_1y'].grid(row=row+2, column=1)

        frame_dict['va_x'] =tk.StringVar()
        frame_dict['amp_x'] = ttk.Entry(frame_dict['checkbox_frame'], width=7, textvariable=frame_dict['va_x'])
        frame_dict['amp_x'].insert(0, '10')
        frame_dict['amp_x'].grid(row=row+3, column=0)

        frame_dict['va_y'] =tk.StringVar()
        frame_dict['amp_y'] = ttk.Entry(frame_dict['checkbox_frame'], width=7, textvariable=frame_dict['va_y'])
        frame_dict['amp_y'].insert(0, '10')
        frame_dict['amp_y'].grid(row=row+3, column=1)

        frame_dict['label_2x'] = tk.Label(frame_dict['checkbox_frame'], text='Rep x')
        frame_dict['label_2x'].grid(row=row+4, column=0)

        frame_dict['label_2y'] = tk.Label(frame_dict['checkbox_frame'], text='Rep y')
        frame_dict['label_2y'].grid(row=row+4, column=1)

        frame_dict['vr_x'] =tk.StringVar()
        frame_dict['rep_x'] = ttk.Entry(frame_dict['checkbox_frame'], width=7, textvariable=frame_dict['vr_x'])
        frame_dict['rep_x'].insert(0, '10')
        frame_dict['rep_x'].grid(row=row+5, column=0)

        frame_dict['vr_y'] =tk.StringVar()
        frame_dict['rep_y'] = ttk.Entry(frame_dict['checkbox_frame'], width=7, textvariable=frame_dict['vr_y'])
        frame_dict['rep_y'].insert(0, '10')
        frame_dict['rep_y'].grid(row=row+5, column=1)

        frame_dict['button'] = ttk.Button(frame_dict['frame'], text='Execute!')
        frame_dict['button'].grid(pady=5)

        for child in frame_dict.values():
            if hasattr(child, 'grid_configure'):
                child.grid_configure(padx=10, pady=5)

    def create_motor_block(self, motor, column):
        '''
        Creates a motor gui element with it's buttons and status
        '''
        # setup variables
        self.motor_status_var[motor] = tk.StringVar()

        self.motor_button[motor] = OrderedDict()

        self.motor_on[motor] = True

        self.motor_button[motor]['on'] = ttk.Button(self.on_off_button,
            text='Motor '+motor, command=lambda: self.motor_button_cmd(motor))
        self.motor_button[motor]['on'].grid(column=column, row=0)

        # setup widgets
        self.motor_block[motor] = ttk.Frame(self.motor_view,
            relief=tk.SUNKEN)
        self.motor_block[motor].grid(column=column, row=1)

        self.motor_label[motor] = tk.Label(self.motor_block[motor],
            text='Motor '+motor)
        self.motor_label[motor].grid()

        self.move_entry[motor] = ttk.Entry(self.motor_block[motor], width=10)
        self.move_entry[motor].grid()

        self.motor_button[motor]['move'] = ttk.Button(self.motor_block[motor],
            text='move', command=lambda: self.move_button_cmd(motor))
        self.motor_button[motor]['move'].grid()

        self.motor_button[motor]['release'] = ttk.Button(self.motor_block[motor],
            text='release', command=lambda: self.release_button_cmd(motor))
        self.motor_button[motor]['release'].grid()

        self.motor_button[motor]['lock'] = ttk.Button(self.motor_block[motor],
            text='lock', command=lambda: self.lock_button_cmd(motor))
        self.motor_button[motor]['lock'].grid()

        self.motor_button[motor]['reset'] = ttk.Button(self.motor_block[motor],
            text='reset', command=lambda: self.reset_button_cmd(motor))
        self.motor_button[motor]['reset'].grid()

        self.motor_button[motor]['zero'] = ttk.Button(self.motor_block[motor],
            text='zero', command=lambda: self.zero_button_cmd(motor))
        self.motor_button[motor]['zero'].grid()

        self.motor_status[motor] = tk.Label(self.motor_block[motor],
            textvariable=self.motor_status_var[motor], width=self.B_WIDTH)
        self.motor_status[motor].grid()

        for child in self.motor_block[motor].winfo_children():
            child.grid_configure(padx=5, pady=5)

        self.motor_button_cmd(motor)

    def motor_button_cmd(self, motor):
        '''
        Action for the move button
        '''
        if self.motor_on[motor]:
            self.motor_block[motor].grid_remove()
            if motor in self.motor_object:
                self.motor_object[motor].cleanup()
                del self.motor_object[motor]
        else:
            self.motor_block[motor].grid()
            self.motor_object[motor] = stp.Motor(self.PINS[motor])
            self.update_message(motor)

        self.motor_on[motor] = not self.motor_on[motor]

    def move_button_cmd(self, motor):
        '''
        Action for the move button
        '''
        steps = self.move_entry[motor].get()
        try:
            steps = int(steps)
        except:
            return
        self.motor_object[motor].move(steps)
        self.update_message(motor)

    def release_button_cmd(self, motor):
        '''
        Action for the move button
        '''
        self.motor_object[motor].release()
        self.update_message(motor)

    def lock_button_cmd(self, motor):
        '''
        Action for the move button
        '''
        self.motor_object[motor].lock()
        self.update_message(motor)

    def reset_button_cmd(self, motor):
        '''
        Action for the move button
        '''
        self.motor_object[motor].reset()
        self.update_message(motor)

    def zero_button_cmd(self, motor):
        '''
        Action for the move button
        '''
        self.motor_object[motor].zero()
        self.update_message(motor)

    def execute_zig_zag_cmd(self):
        mx = 'm'+str(self.zig_zag_frame['checkbox']['var_x'].get())
        my = 'm'+str(self.zig_zag_frame['checkbox']['var_y'].get())
        ax = int(self.zig_zag_frame['amp_x'].get())
        ay = int(self.zig_zag_frame['amp_y'].get())
        rx = int(self.zig_zag_frame['rep_x'].get())
        ry = int(self.zig_zag_frame['rep_y'].get())
        if mx == my :
            self.dialog_window('error', 'You can\'t control two motors in the same port')
            return
        else:
            motors = (stp.Motor(self.PINS[mx]), stp.Motor(self.PINS[my]))
            stp.zig_zag(*motors, (ax, rx), (ay, ry))
            for motor in motors:
                motor.cleanup()
                del motor

    def execute_s_spiral_cmd(self):
        mx = 'm'+str(self.s_spiral_frame['checkbox']['var_x'].get())
        my = 'm'+str(self.s_spiral_frame['checkbox']['var_y'].get())
        ax = int(self.s_spiral_frame['amp_x'].get())
        rx = int(self.s_spiral_frame['rep_x'].get())
        if mx == my :
            self.dialog_window('error', 'You can\'t control two motors in the same port')
            return
        else:
            motors = (stp.Motor(self.PINS[mx]), stp.Motor(self.PINS[my]))
            stp.square_spiral(*motors, (ax, rx))
            for motor in motors:
                motor.cleanup()
                del motor


RPistepper/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
RPistepper is a library containing:
    * A class to control a stepper motor with a RPi.
    * A function to execute a zig-zag motion with two motors.
    * A function to execute a square_spiral motion with two motors.

Copyright (C) 2015 Luiz Eduardo Amaral <luizamaral306@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

#______________________________________________________________________
# imports
import RPi.GPIO as GPIO
from time import sleep

#______________________________________________________________________
# version
__version__ = '0.3a0'

#______________________________________________________________________
# globals
m0 = [2, 3, 4, 17],
m1 = [14, 15, 18, 23],
m2 = [27, 22, 10, 9],
m3 = [24, 25, 8, 7],
m4 = [11, 0, 5, 6],
m5 = [1, 12, 16, 20],
m6 = [13, 19, 26, 21]

#______________________________________________________________________
# classes
class Motor(object):
    '''
    This class allows the user to control a 6 pin stepper motor using
    4 GPIO pins of a RPi.

    Software uses BCM mode for pin indexing.

    This class is best used with the 'with' statement to properly
    handle the cleanup of the GPIOs.

    self.steps is a property of this class that will get the number of
    steps taken from the initial position or set to a specific step,
    similar to self.move.

    In order to save power, it's advised to call self.release() when
    the motor is idle.
    '''

    #__________________________________________________________________
    # class attributes
    DELAY = 0.02
    VERBOSE = False

    #__________________________________________________________________
    # magic methods
    def __init__(self, pins, delay=DELAY, verbose=VERBOSE):
        '''
        Arguments are a list with the 4 pins (Coil_A1, Coil_A2,
        Coil_B1, Coil_B2), the delay between steps (default = 20ms) and
        verbose to display reports on the motor movements, the last two
        are optional.
        '''
        self.PINS = pins
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(self.PINS, GPIO.OUT)
        self.DELAY = delay
        self.VERBOSE = verbose
        self.actual_state = []
        self.locked = False
        self._step_list = [
            (1, 1, 0, 0),
            (1, 0, 1, 0),
            (0, 1, 1, 0),
            (0, 1, 0, 1),
            (0, 0, 1, 1),
            (1, 0, 0, 1)]
        self._release_all = [0]*4
        self._steps = 0
        self._set_step(self._step_list[0])
        self.release()

    def __repr__(self):
        return 'Motor at pins: {0}, Steps: {1}, Position: {2}'.format(
            self.PINS, self.steps, self.actual_state)

    def __enter__(self):
        return self

    def __exit__(self, type, value, tb):
        self.cleanup()

    #__________________________________________________________________
    # properties
    @property
    def steps(self):
        '''
        Number of steps taken from the initial position
        '''
        return self._steps

    @steps.setter
    def steps(self, value):
        steps = value - self._steps
        self.move(steps)

    #__________________________________________________________________
    # methods
    def move(self, steps):
        '''
        Moves the motor 'steps' steps. Negative steps moves the motor
        backwards
        '''
        if steps == 0:
            return
        if self.VERBOSE:
            print(str(self)+ ', Moving: {0} steps'.format(steps))
        rotation = steps//abs(steps)
        for i in range(0, steps, rotation):
            index = (self._steps + rotation)%len(self._step_list)
            self._set_step(self._step_list[index])
            sleep(self.DELAY)
            self._steps += rotation
        self.locked = True

    def release(self):
        '''
        Sets all pins low. Power saving mode
        '''
        self._set_step(self._release_all)
        self.actual_state = self._release_all
        self.locked = False

    def lock(self):
        '''
        Locks the pins in the last known step
        '''
        index = (self._steps)%len(self._step_list)
        self._set_step(self._step_list[index])
        self.locked = True

    def reset(self):
        '''
        Returns the motor to it's initial position
        '''
        self.steps = 0
        self.locked = True

    def zero(self):
        '''
        Sets the motor to the next position which Coil_A1 and Coil_A2
        are on. Sets this position as the reference (steps = 0).
        '''
        self._steps = 0
        self.steps = 6
        self.steps = 0
        self.locked = True

    def cleanup(self):
        '''
        Cleans the GPIO resources
        '''
        GPIO.cleanup(self.PINS)

    #__________________________________________________________________
    # private methods
    def _set_step(self, states):
        '''
        Sets the pins A_1, A_2, B_1, B_2, 1 or 0 (HIGH ou LOW)
        '''
        self.actual_state = states
        GPIO.output(self.PINS, states)

#______________________________________________________________________
# backwards compatibility
RPiStepper = Motor

#______________________________________________________________________
# functions
def zig_zag(motor1, motor2, amp1, amp2, delay=None):
    '''
    Executes a zig-zag movement with two RPistepper objects.
    Arguments are: motor1 and motor2 objects and amp1, amp2, the
    amplitude of movement, a tuple (step, rep) representing the number
    of steps per iteration and the number of iterations of the following
    algorithm:
        Repeat rep1 times:
            1. Moves motor 2 step2*rep2 steps forward
            2. Moves motor 1 step1 steps forward
            3. Moves motor 2 step2*rep2 steps backwards
            4. Moves motor 1 step1 steps forward
        Reset to initial state
        Release the motors
    It's possible to change the delay between steps with the 'delay'
    argument
    '''
    step1, rep1 = amp1
    step2, rep2 = amp2
    if delay:
        motor1.delay = delay
        motor2.delay = delay
    for i in range(rep1):
        motor2.move(step2*rep2)
        motor1.move(step1)
        motor2.move(-step2*rep2)
        motor1.move(step1)
    motor1.reset()
    motor2.reset()
    motor1.release()
    motor2.release()

def square_spiral(motor1, motor2, amplitude, delay=None):
    '''
    Executes a square spiral movement with two RPistepper objects.
    Arguments are: motor1 and motor2 objects and the amplitude of movement,
    a tuple (step, rep) representing the number of steps per iteration and
    the number of iterations of the following algorithm:
        for i in range(rep):
            1. Moves motor 2 to position i
            2. Moves motor 1 to position i
            3. Moves motor 1 to position -i
            4. Moves motor 2 to position -i
        Reset to initial state
        Release the motors
    It's possible to change the delay between steps with the 'delay'
    argument
    '''
    step, rep = amplitude
    if delay:
        motor1.delay = delay
        motor2.delay = delay
    for i in range(rep):
        motor2.steps = (i+1)*step
        motor1.steps = (i+1)*step
        motor1.steps = -(i+1)*step
        motor2.steps = -(i+1)*step
    motor1.reset()
    motor2.reset()
    motor1.release()
    motor2.release()
